### PR TITLE
Use go-tc instead of shell

### PR DESCRIPTION
**Pull request by kebe7jun**
```console
istio-proxy@sleep-784d69f945-bjb4m:/$ tc filter show dev eth0 ingress
filter protocol all pref 66 bpf chain 0
filter protocol all pref 66 bpf chain 0 handle 0x1 mb_tc_ingress direct-action not_in_hw id 1236 tag 8228568dfe62b337
istio-proxy@sleep-784d69f945-bjb4m:/$ tc filter show dev eth0 egress
filter protocol all pref 66 bpf chain 0
filter protocol all pref 66 bpf chain 0 handle 0x1 mb_tc_egress direct-action not_in_hw id 1275 tag bceb4eff4a7e13e6
```

```console
# bpftool prog
1242: cgroup_sock_addr  name mb_sock_connect  tag 899fc03b5f58d07f  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 9512B  jited 6197B  memlock 12288B  map_ids 1553,1550,1551,1552,1558
	btf_id 59496
1247: sock_ops  name mb_sockops  tag b2c6c93d04f557a3  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 1600B  jited 879B  memlock 4096B  map_ids 1550,1552,1560,1559
	btf_id 59500
1252: cgroup_sockopt  name mb_get_sockopt  tag ffbd8894d2844c48  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 968B  jited 554B  memlock 4096B  map_ids 1559
	btf_id 59504
1257: sk_msg  name mb_msg_redir  tag 799589f0f98c291f  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 504B  jited 297B  memlock 4096B  map_ids 1560
	btf_id 59508
1262: cgroup_sock_addr  name mb_bind  tag 57cd311f2e27366b  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 16B  jited 18B  memlock 4096B
	btf_id 59512
1267: cgroup_sock_addr  name mb_sendmsg4  tag 722e3d7b05977148  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 800B  jited 466B  memlock 4096B  map_ids 1550
	btf_id 59516
1272: cgroup_sock_addr  name mb_recvmsg4  tag 736ee1370a31bbd0  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 864B  jited 500B  memlock 4096B  map_ids 1550
	btf_id 59520
1274: sched_cls  name mb_tc_ingress  tag 8228568dfe62b337  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 3016B  jited 1954B  memlock 4096B  map_ids 1559,1551
	btf_id 59523
1275: sched_cls  name mb_tc_egress  tag bceb4eff4a7e13e6  gpl
	loaded_at 2022-09-23T10:23:37+0800  uid 0
	xlated 1504B  jited 885B  memlock 4096B  map_ids 1559
	btf_id 59523
```

### CODE REVIEW

**Comment by kebe7jun** [comment]
@dddddai PTAL.

File: internal/cni-server/cni-plugin.go [hunk 1]
```diff
@@ -319,38 +325,126 @@ func skipListening(serviceMeshMode string, pid string) bool {
...
+	if err := rtnl.Filter().Add(&filter); err != nil {
+		return err
 	}
 
 	s.Lock()
 	s.qdiscs[inode] = qdisc{
 		netns:     netns,
 		device:    dev,
-		hasClsact: hasCls,
+		hasClsact: !find,
```
Comment (dddddai) [inline internal/cni-server/cni-plugin.go:447]: this should be `find`?

File: internal/cni-server/cni-plugin.go [hunk 2]
```diff
@@ -366,25 +460,62 @@ func (s *server) cleanUpTC() {
...
+				if err := rtnl.Close(); err != nil {
+					log.Errorf("could not close rtnetlink socket: %v\n", err)
 				}
-				cmd = exec.Command("sh", "-c", fmt.Sprintf("tc filter delete dev %s ingress prio 66", q.device))
-				err = cmd.Run()
-				if code := cmd.ProcessState.ExitCode(); code != 0 || err != nil {
-					return fmt.Errorf("failed to delete ingress filter from %s, unexpected exit code: %d, err: %v", q.device, code, err)
+			}()
+			if q.hasClsact {
+				err := rtnl.Qdisc().Delete(&tc.Object{
```
Comment (dddddai) [inline internal/cni-server/cni-plugin.go:477]: this should be `!q.hasClsact`?

**Review by dddddai — COMMENTED** [review]
Nice!

File: internal/cni-server/cni-plugin.go (hunk 1 above)
Comment (kebe7jun) [reply internal/cni-server/cni-plugin.go:447]: find means clsact already exists before we create, it was created by others.
hasClsact means that this clsact was created by us.
may be we should change `hasClsact` to `managedClsact` is better.

File: internal/cni-server/cni-plugin.go [hunk 3]
```diff
@@ -267,7 +269,11 @@ func (s *server) checkAndRepairPodPrograms() error {
 				ifaces, _ := net.Interfaces()
 				for _, iface := range ifaces {
 					if (iface.Flags&net.FlagLoopback) == 0 && (iface.Flags&net.FlagUp) != 0 {
-						return s.attachTC(netns.Path(), iface.Name)
+						err := s.attachTC(netns.Path(), iface.Name)
```
Comment (dddddai) [inline internal/cni-server/cni-plugin.go:272]: I think we can pass `iface/iface.Index` here

File: internal/cni-server/cni-plugin.go [hunk 4]
```diff
@@ -319,38 +325,126 @@ func skipListening(serviceMeshMode string, pid string) bool {
...
+		return err
 	}
 
 	s.Lock()
 	s.qdiscs[inode] = qdisc{
-		netns:     netns,
-		device:    dev,
-		hasClsact: hasCls,
+		netns:         netns,
+		device:        dev,
```
Comment (dddddai) [inline internal/cni-server/cni-plugin.go:446]: use `iface.Index` instead?

File: internal/ebpfs/prog.go [hunk 5]
```diff
@@ -69,3 +72,52 @@ func UnLoadMBProgs() error {
...
+	coll, err := ebpf.LoadCollectionSpec("bpf/mb_tc.o")
+	if err != nil {
+		return err
+	}
+	type progs struct {
+		Ingress *ebpf.Program `ebpf:"mb_tc_ingress"`
+		Egress  *ebpf.Program `ebpf:"mb_tc_egress"`
+	}
+	ps := progs{}
+	err = coll.LoadAndAssign(&ps, &ebpf.CollectionOptions{
```
Comment (dddddai) [inline internal/ebpfs/prog.go:111]: One question: how can we tell which section to load?

Comment (kebe7jun) [reply internal/ebpfs/prog.go:111]: This does not use `sec` to distinguish, it uses function names as separate procedures, see the description. `mb_tc_ingress` or `mb_tc_egress`.

File: internal/cni-server/cni-plugin.go [hunk 6]
```diff
@@ -319,38 +325,126 @@ func skipListening(serviceMeshMode string, pid string) bool {
...
+				Flags: uint32Ptr(0x1),
+			},
+		},
+	}
+	if err := rtnl.Filter().Add(&filter); err != nil {
+		return err
+	}
+	egress := ebpfs.GetTCEgressProg()
+	if ing == nil {
+		return fmt.Errorf("can not get ingress prog")
```
Comment (dddddai) [inline internal/cni-server/cni-plugin.go:416]: ingress -> egress

**Review by dddddai — APPROVED** [review]

**Pull request merged**